torch/csrc/autograd/functions/accumulate_grad.h

```diff
@@ -110,9 +171,33 @@ struct TORCH_API AccumulateGrad : public Node {
         // place. `variable_grad` is thus still referring to the same tensor
         // after the operation.
         variable_grad += new_grad;
+        // ^ We could enforce the contract more aggressively here by writing:
+        // if (variable_grad.is_sparse() || new_grad.is_sparse()) {
+        //   variable_grad += new_grad;
+        // } else if (obeys_layout_contract(variable_grad, variable)) {
+        //   variable_grad += new_grad;
+        // } else {
+        //   result = at::empty_strided(variable.sizes(), variable.strides(),
+        //                              variable.options().memory_format(c10::nullopt));
+        //   update_grad(at::native::add_out(result, variable_grad, new_grad, 1.0);
+        // }
+        // However, that accumulation is sometimes in place and sometimes not,
```

Comment:
Yes for BC reasons we should be very careful about doing this!

Reply:
As currently written, the PR's decisions about whether grads are accumulated in place or out of place is fully BC (identical to master).  The comment here is a tradeoff we could make:  sacrifice fully BC in-place behavior for more rigid param<->grad layout matching.

Reply:
Yes, the PR doesn't go down that road, and I think it's a good thing !

Reply:
SGTM, i'll keep it as written.